// vscode-azureappservice/lib/tree/AppSettingsTreeItem.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const extensionVariables_1 = require("../extensionVariables");
const AppSettingTreeItem_1 = require("./AppSettingTreeItem");
class AppSettingsTreeItem {
    constructor(client) {
        this.label = 'Application Settings';
        this.childTypeLabel = 'App Setting';
        this.contextValue = AppSettingsTreeItem.contextValue;
        this._client = client;
    }
    get id() {
        return 'application';
    }
    get iconPath() {
        return {
            light: path.join(__filename, '..', '..', '..', 'resources', 'light', 'AppSettings_color.svg'),
            dark: path.join(__filename, '..', '..', '..', 'resources', 'dark', 'AppSettings_color.svg')
        };
    }
    hasMoreChildren() {
        return false;
    }
    loadMoreChildren() {
        return __awaiter(this, void 0, void 0, function* () {
            this._settings = yield this._client.listApplicationSettings();
            const treeItems = [];
            Object.keys(this._settings.properties).forEach((key) => {
                treeItems.push(new AppSettingTreeItem_1.AppSettingTreeItem(key, this._settings.properties[key]));
            });
            return treeItems;
        });
    }
    editSettingItem(oldKey, newKey, value) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this._settings.properties) {
                if (oldKey !== newKey) {
                    delete this._settings.properties[oldKey];
                }
                this._settings.properties[newKey] = value;
            }
            yield this._client.updateApplicationSettings(this._settings);
        });
    }
    deleteSettingItem(key) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this._settings.properties) {
                delete this._settings.properties[key];
            }
            yield this._client.updateApplicationSettings(this._settings);
        });
    }
    createChild(_node, showCreatingNode) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._settings) {
                yield this.loadMoreChildren();
            }
            const newKey = yield extensionVariables_1.ext.ui.showInputBox({
                prompt: 'Enter new setting key',
                validateInput: (v) => this.validateNewKeyInput(v)
            });
            const newValue = yield extensionVariables_1.ext.ui.showInputBox({
                prompt: `Enter setting value for "${newKey}"`
            });
            if (!this._settings.properties) {
                this._settings.properties = {};
            }
            showCreatingNode(newKey);
            this._settings.properties[newKey] = newValue;
            yield this._client.updateApplicationSettings(this._settings);
            return new AppSettingTreeItem_1.AppSettingTreeItem(newKey, newValue);
        });
    }
    validateNewKeyInput(newKey, oldKey) {
        newKey = newKey ? newKey.trim() : '';
        oldKey = oldKey ? oldKey.trim().toLowerCase() : oldKey;
        if (newKey.length === 0) {
            return 'Key must have at least one non-whitespace character.';
        }
        if (this._settings.properties && newKey.toLowerCase() !== oldKey) {
            for (const key of Object.keys(this._settings.properties)) {
                if (key.toLowerCase() === newKey.toLowerCase()) {
                    return `Setting "${newKey}" already exists.`;
                }
            }
        }
        return undefined;
    }
}
AppSettingsTreeItem.contextValue = 'applicationSettings';
exports.AppSettingsTreeItem = AppSettingsTreeItem;
