// vscode-azurekudu/lib/models/commandResult.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * Class representing a CommandResult.
 */
class CommandResult {
  /**
   * Create a CommandResult.
   * @member {string} [output]
   * @member {string} [error]
   * @member {number} [exitCode]
   */
  constructor() {
  }

  /**
   * Defines the metadata of CommandResult
   *
   * @returns {object} metadata of CommandResult
   *
   */
  mapper() {
    return {
      required: false,
      serializedName: 'CommandResult',
      type: {
        name: 'Composite',
        className: 'CommandResult',
        modelProperties: {
          output: {
            required: false,
            serializedName: 'Output',
            type: {
              name: 'String'
            }
          },
          error: {
            required: false,
            serializedName: 'Error',
            type: {
              name: 'String'
            }
          },
          exitCode: {
            required: false,
            serializedName: 'ExitCode',
            type: {
              name: 'Number'
            }
          }
        }
      }
    };
  }
}

module.exports = CommandResult;
