// vscode-azurekudu/lib/models/logEntry.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * Class representing a LogEntry.
 */
class LogEntry {
  /**
   * Create a LogEntry.
   * @member {date} [logTime]
   * @member {string} [id]
   * @member {string} [message]
   * @member {number} [type]
   * @member {string} [detailsUrl]
   */
  constructor() {
  }

  /**
   * Defines the metadata of LogEntry
   *
   * @returns {object} metadata of LogEntry
   *
   */
  mapper() {
    return {
      required: false,
      serializedName: 'LogEntry',
      type: {
        name: 'Composite',
        className: 'LogEntry',
        modelProperties: {
          logTime: {
            required: false,
            serializedName: 'log_time',
            type: {
              name: 'DateTime'
            }
          },
          id: {
            required: false,
            serializedName: 'id',
            type: {
              name: 'String'
            }
          },
          message: {
            required: false,
            serializedName: 'message',
            type: {
              name: 'String'
            }
          },
          type: {
            required: false,
            serializedName: 'type',
            type: {
              name: 'Number'
            }
          },
          detailsUrl: {
            required: false,
            serializedName: 'details_url',
            type: {
              name: 'String'
            }
          }
        }
      }
    };
  }
}

module.exports = LogEntry;
