// vscode-azurekudu/lib/models/repositoryInfo.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * Class representing a RepositoryInfo.
 */
class RepositoryInfo {
  /**
   * Create a RepositoryInfo.
   * @member {number} [type]
   * @member {string} [gitUrl]
   */
  constructor() {
  }

  /**
   * Defines the metadata of RepositoryInfo
   *
   * @returns {object} metadata of RepositoryInfo
   *
   */
  mapper() {
    return {
      required: false,
      serializedName: 'RepositoryInfo',
      type: {
        name: 'Composite',
        className: 'RepositoryInfo',
        modelProperties: {
          type: {
            required: false,
            serializedName: 'Type',
            type: {
              name: 'Number'
            }
          },
          gitUrl: {
            required: false,
            serializedName: 'GitUrl',
            type: {
              name: 'String'
            }
          }
        }
      }
    };
  }
}

module.exports = RepositoryInfo;
