// vscode-azurekudu/lib/models/siteExtensionInfo.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * Class representing a SiteExtensionInfo.
 */
class SiteExtensionInfo {
  /**
   * Create a SiteExtensionInfo.
   * @member {string} [id]
   * @member {string} [title]
   * @member {number} [type]
   * @member {string} [summary]
   * @member {string} [description]
   * @member {string} [version]
   * @member {string} [extensionUrl]
   * @member {string} [projectUrl]
   * @member {string} [iconUrl]
   * @member {string} [licenseUrl]
   * @member {string} [feedUrl]
   * @member {array} [authors]
   * @member {string} [installerCommandLineParams]
   * @member {date} [publishedDateTime]
   * @member {number} [downloadCount]
   * @member {boolean} [localIsLatestVersion]
   * @member {string} [localPath]
   * @member {date} [installedDateTime]
   * @member {string} [provisioningState]
   * @member {string} [comment]
   */
  constructor() {
  }

  /**
   * Defines the metadata of SiteExtensionInfo
   *
   * @returns {object} metadata of SiteExtensionInfo
   *
   */
  mapper() {
    return {
      required: false,
      serializedName: 'SiteExtensionInfo',
      type: {
        name: 'Composite',
        className: 'SiteExtensionInfo',
        modelProperties: {
          id: {
            required: false,
            serializedName: 'id',
            type: {
              name: 'String'
            }
          },
          title: {
            required: false,
            serializedName: 'title',
            type: {
              name: 'String'
            }
          },
          type: {
            required: false,
            serializedName: 'type',
            type: {
              name: 'Number'
            }
          },
          summary: {
            required: false,
            serializedName: 'summary',
            type: {
              name: 'String'
            }
          },
          description: {
            required: false,
            serializedName: 'description',
            type: {
              name: 'String'
            }
          },
          version: {
            required: false,
            serializedName: 'version',
            type: {
              name: 'String'
            }
          },
          extensionUrl: {
            required: false,
            serializedName: 'extension_url',
            type: {
              name: 'String'
            }
          },
          projectUrl: {
            required: false,
            serializedName: 'project_url',
            type: {
              name: 'String'
            }
          },
          iconUrl: {
            required: false,
            serializedName: 'icon_url',
            type: {
              name: 'String'
            }
          },
          licenseUrl: {
            required: false,
            serializedName: 'license_url',
            type: {
              name: 'String'
            }
          },
          feedUrl: {
            required: false,
            serializedName: 'feed_url',
            type: {
              name: 'String'
            }
          },
          authors: {
            required: false,
            serializedName: 'authors',
            type: {
              name: 'Sequence',
              element: {
                  required: false,
                  serializedName: 'StringElementType',
                  type: {
                    name: 'String'
                  }
              }
            }
          },
          installerCommandLineParams: {
            required: false,
            serializedName: 'installer_command_line_params',
            type: {
              name: 'String'
            }
          },
          publishedDateTime: {
            required: false,
            serializedName: 'published_date_time',
            type: {
              name: 'DateTime'
            }
          },
          downloadCount: {
            required: false,
            serializedName: 'download_count',
            type: {
              name: 'Number'
            }
          },
          localIsLatestVersion: {
            required: false,
            serializedName: 'local_is_latest_version',
            type: {
              name: 'Boolean'
            }
          },
          localPath: {
            required: false,
            serializedName: 'local_path',
            type: {
              name: 'String'
            }
          },
          installedDateTime: {
            required: false,
            serializedName: 'installed_date_time',
            type: {
              name: 'DateTime'
            }
          },
          provisioningState: {
            required: false,
            serializedName: 'provisioningState',
            type: {
              name: 'String'
            }
          },
          comment: {
            required: false,
            serializedName: 'comment',
            type: {
              name: 'String'
            }
          }
        }
      }
    };
  }
}

module.exports = SiteExtensionInfo;
