// vscode-azurekudu/lib/kuduClient.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

/* jshint latedef:false */
/* jshint forin:false */
/* jshint noempty:false */

'use strict';

const msRest = require('ms-rest');
const ServiceClient = msRest.ServiceClient;

const models = require('./models');
const operations = require('./operations');


/** Class representing a KuduClient. */
class KuduClient extends ServiceClient {
  /**
   * Create a KuduClient.
   * @param {credentials} credentials - Subscription credentials which uniquely identify client subscription.
   * @param {string} [baseUri] - The base URI of the service.
   * @param {object} [options] - The parameter options
   * @param {Array} [options.filters] - Filters to be added to the request pipeline
   * @param {object} [options.requestOptions] - Options for the underlying request object
   * {@link https://github.com/request/request#requestoptions-callback Options doc}
   * @param {boolean} [options.noRetryPolicy] - If set to true, turn off default retry policy
   */
  constructor(credentials, baseUri, options) {
    if (credentials === null || credentials === undefined) {
      throw new Error('\'credentials\' cannot be null.');
    }

    if (!options) options = {};

    super(credentials, options);

    this.baseUri = baseUri;
    if (!this.baseUri) {
      this.baseUri = 'https://scm.azurewebsites.net';
    }
    this.credentials = credentials;

    let packageInfo = this.getPackageJsonInfo(__dirname);
    this.addUserAgentInfo(`${packageInfo.name}/${packageInfo.version}`);
    this.command = new operations.Command(this);
    this.deployment = new operations.Deployment(this);
    this.diagnostics = new operations.Diagnostics(this);
    this.environment = new operations.Environment(this);
    this.error404 = new operations.Error404(this);
    this.functionModel = new operations.FunctionModel(this);
    this.infoRefs = new operations.InfoRefs(this);
    this.jobs = new operations.Jobs(this);
    this.liveScm = new operations.LiveScm(this);
    this.liveScmEditor = new operations.LiveScmEditor(this);
    this.process = new operations.Process(this);
    this.pushDeployment = new operations.PushDeployment(this);
    this.runtime = new operations.Runtime(this);
    this.settings = new operations.Settings(this);
    this.siteExtension = new operations.SiteExtension(this);
    this.sSHKey = new operations.SSHKey(this);
    this.vfs = new operations.Vfs(this);
    this.webHooks = new operations.WebHooks(this);
    this.zip = new operations.Zip(this);
    this.models = models;
    msRest.addSerializationMixin(this);
  }

}

module.exports = KuduClient;
module.exports['default'] = KuduClient;
module.exports.KuduClient = KuduClient;
module.exports.KuduModels = models;
