// vscode-azurekudu/lib/operations/index.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

/* jshint latedef:false */
/* jshint forin:false */
/* jshint noempty:false */

'use strict';

exports.Command = require('./command');
exports.Deployment = require('./deployment');
exports.Diagnostics = require('./diagnostics');
exports.Environment = require('./environment');
exports.Error404 = require('./error404');
exports.FunctionModel = require('./functionModel');
exports.InfoRefs = require('./infoRefs');
exports.Jobs = require('./jobs');
exports.LiveScm = require('./liveScm');
exports.LiveScmEditor = require('./liveScmEditor');
exports.Process = require('./process');
exports.PushDeployment = require('./pushDeployment');
exports.Runtime = require('./runtime');
exports.Settings = require('./settings');
exports.SiteExtension = require('./siteExtension');
exports.SSHKey = require('./sSHKey');
exports.Vfs = require('./vfs');
exports.WebHooks = require('./webHooks');
exports.Zip = require('./zip');
