// vscode-azurekudu/lib/models/commandInput.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * Class representing a CommandInput.
 */
class CommandInput {
  /**
   * Create a CommandInput.
   * @member {string} [command]
   * @member {string} [dir]
   */
  constructor() {
  }

  /**
   * Defines the metadata of CommandInput
   *
   * @returns {object} metadata of CommandInput
   *
   */
  mapper() {
    return {
      required: false,
      serializedName: 'CommandInput',
      type: {
        name: 'Composite',
        className: 'CommandInput',
        modelProperties: {
          command: {
            required: false,
            serializedName: 'command',
            type: {
              name: 'String'
            }
          },
          dir: {
            required: false,
            serializedName: 'dir',
            type: {
              name: 'String'
            }
          }
        }
      }
    };
  }
}

module.exports = CommandInput;
