// vscode-azurekudu/lib/models/diagnosticsSettings.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * Class representing a DiagnosticsSettings.
 */
class DiagnosticsSettings {
  /**
   * Create a DiagnosticsSettings.
   * @member {boolean} [azureDriveEnabled]
   * @member {number} [azureDriveTraceLevel]
   * @member {boolean} [azureTableEnabled]
   * @member {number} [azureTableTraceLevel]
   * @member {boolean} [azureBlobEnabled]
   * @member {number} [azureBlobTraceLevel]
   */
  constructor() {
  }

  /**
   * Defines the metadata of DiagnosticsSettings
   *
   * @returns {object} metadata of DiagnosticsSettings
   *
   */
  mapper() {
    return {
      required: false,
      serializedName: 'DiagnosticsSettings',
      type: {
        name: 'Composite',
        className: 'DiagnosticsSettings',
        modelProperties: {
          azureDriveEnabled: {
            required: false,
            serializedName: 'AzureDriveEnabled',
            type: {
              name: 'Boolean'
            }
          },
          azureDriveTraceLevel: {
            required: false,
            serializedName: 'AzureDriveTraceLevel',
            type: {
              name: 'Number'
            }
          },
          azureTableEnabled: {
            required: false,
            serializedName: 'AzureTableEnabled',
            type: {
              name: 'Boolean'
            }
          },
          azureTableTraceLevel: {
            required: false,
            serializedName: 'AzureTableTraceLevel',
            type: {
              name: 'Number'
            }
          },
          azureBlobEnabled: {
            required: false,
            serializedName: 'AzureBlobEnabled',
            type: {
              name: 'Boolean'
            }
          },
          azureBlobTraceLevel: {
            required: false,
            serializedName: 'AzureBlobTraceLevel',
            type: {
              name: 'Number'
            }
          }
        }
      }
    };
  }
}

module.exports = DiagnosticsSettings;
