// vscode-azurekudu/lib/models/functionEnvelope.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * Class representing a FunctionEnvelope.
 */
class FunctionEnvelope {
  /**
   * Create a FunctionEnvelope.
   * @member {string} [name]
   * @member {string} [functionAppId]
   * @member {string} [scriptRootPathHref]
   * @member {string} [scriptHref]
   * @member {string} [configHref]
   * @member {string} [secretsFileHref]
   * @member {string} [href]
   * @member {object} [config]
   * @member {object} [files]
   * @member {string} [testData]
   */
  constructor() {
  }

  /**
   * Defines the metadata of FunctionEnvelope
   *
   * @returns {object} metadata of FunctionEnvelope
   *
   */
  mapper() {
    return {
      required: false,
      serializedName: 'FunctionEnvelope',
      type: {
        name: 'Composite',
        className: 'FunctionEnvelope',
        modelProperties: {
          name: {
            required: false,
            serializedName: 'name',
            type: {
              name: 'String'
            }
          },
          functionAppId: {
            required: false,
            serializedName: 'function_app_id',
            type: {
              name: 'String'
            }
          },
          scriptRootPathHref: {
            required: false,
            serializedName: 'script_root_path_href',
            type: {
              name: 'String'
            }
          },
          scriptHref: {
            required: false,
            serializedName: 'script_href',
            type: {
              name: 'String'
            }
          },
          configHref: {
            required: false,
            serializedName: 'config_href',
            type: {
              name: 'String'
            }
          },
          secretsFileHref: {
            required: false,
            serializedName: 'secrets_file_href',
            type: {
              name: 'String'
            }
          },
          href: {
            required: false,
            serializedName: 'href',
            type: {
              name: 'String'
            }
          },
          config: {
            required: false,
            serializedName: 'config',
            type: {
              name: 'Object'
            }
          },
          files: {
            required: false,
            serializedName: 'files',
            type: {
              name: 'Dictionary',
              value: {
                  required: false,
                  serializedName: 'StringElementType',
                  type: {
                    name: 'String'
                  }
              }
            }
          },
          testData: {
            required: false,
            serializedName: 'test_data',
            type: {
              name: 'String'
            }
          }
        }
      }
    };
  }
}

module.exports = FunctionEnvelope;
