// vscode-azurekudu/lib/models/functionSecrets.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * Class representing a FunctionSecrets.
 */
class FunctionSecrets {
  /**
   * Create a FunctionSecrets.
   * @member {string} [key]
   * @member {string} [triggerUrl]
   */
  constructor() {
  }

  /**
   * Defines the metadata of FunctionSecrets
   *
   * @returns {object} metadata of FunctionSecrets
   *
   */
  mapper() {
    return {
      required: false,
      serializedName: 'FunctionSecrets',
      type: {
        name: 'Composite',
        className: 'FunctionSecrets',
        modelProperties: {
          key: {
            required: false,
            serializedName: 'key',
            type: {
              name: 'String'
            }
          },
          triggerUrl: {
            required: false,
            serializedName: 'trigger_url',
            type: {
              name: 'String'
            }
          }
        }
      }
    };
  }
}

module.exports = FunctionSecrets;
