// vscode-azurekudu/lib/models/masterKey.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * Class representing a MasterKey.
 */
class MasterKey {
  /**
   * Create a MasterKey.
   * @member {string} [masterKey]
   */
  constructor() {
  }

  /**
   * Defines the metadata of MasterKey
   *
   * @returns {object} metadata of MasterKey
   *
   */
  mapper() {
    return {
      required: false,
      serializedName: 'MasterKey',
      type: {
        name: 'Composite',
        className: 'MasterKey',
        modelProperties: {
          masterKey: {
            required: false,
            serializedName: 'masterKey',
            type: {
              name: 'String'
            }
          }
        }
      }
    };
  }
}

module.exports = MasterKey;
