// vscode-azurekudu/lib/models/runtimeInfo.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * Class representing a RuntimeInfo.
 */
class RuntimeInfo {
  /**
   * Create a RuntimeInfo.
   * @member {array} [nodejs]
   * @member {object} [system]
   */
  constructor() {
  }

  /**
   * Defines the metadata of RuntimeInfo
   *
   * @returns {object} metadata of RuntimeInfo
   *
   */
  mapper() {
    return {
      required: false,
      serializedName: 'RuntimeInfo',
      type: {
        name: 'Composite',
        className: 'RuntimeInfo',
        modelProperties: {
          nodejs: {
            required: false,
            serializedName: 'nodejs',
            type: {
              name: 'Sequence',
              element: {
                  required: false,
                  serializedName: 'ObjectElementType',
                  type: {
                    name: 'Dictionary',
                    value: {
                        required: false,
                        serializedName: 'StringElementType',
                        type: {
                          name: 'String'
                        }
                    }
                  }
              }
            }
          },
          system: {
            required: false,
            serializedName: 'System',
            type: {
              name: 'Object'
            }
          }
        }
      }
    };
  }
}

module.exports = RuntimeInfo;
