// vscode-azurekudu/lib/models/webHook.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * Class representing a WebHook.
 */
class WebHook {
  /**
   * Create a WebHook.
   * @member {string} [id]
   * @member {string} [event]
   * @member {string} [url]
   * @member {boolean} [insecureSsl]
   * @member {string} [lastStatus]
   * @member {string} [lastReason]
   * @member {date} [lastDatetime]
   * @member {string} [lastContext]
   */
  constructor() {
  }

  /**
   * Defines the metadata of WebHook
   *
   * @returns {object} metadata of WebHook
   *
   */
  mapper() {
    return {
      required: false,
      serializedName: 'WebHook',
      type: {
        name: 'Composite',
        className: 'WebHook',
        modelProperties: {
          id: {
            required: false,
            serializedName: 'id',
            type: {
              name: 'String'
            }
          },
          event: {
            required: false,
            serializedName: 'event',
            type: {
              name: 'String'
            }
          },
          url: {
            required: false,
            serializedName: 'url',
            type: {
              name: 'String'
            }
          },
          insecureSsl: {
            required: false,
            serializedName: 'insecure_ssl',
            type: {
              name: 'Boolean'
            }
          },
          lastStatus: {
            required: false,
            serializedName: 'last_status',
            type: {
              name: 'String'
            }
          },
          lastReason: {
            required: false,
            serializedName: 'last_reason',
            type: {
              name: 'String'
            }
          },
          lastDatetime: {
            required: false,
            serializedName: 'last_datetime',
            type: {
              name: 'DateTime'
            }
          },
          lastContext: {
            required: false,
            serializedName: 'last_context',
            type: {
              name: 'String'
            }
          }
        }
      }
    };
  }
}

module.exports = WebHook;
