// vscode-azureappservice/lib/extensionVariables.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
const vscode_azureextensionui_1 = require("vscode-azureextensionui");
const localize_1 = require("./localize");
class UninitializedExtensionVariables {
    constructor() {
        this._error = new Error(localize_1.localize('uninitializedError', '"registerAppServiceExtensionVariables" must be called before using the vscode-azureappservice package.'));
    }
    get outputChannel() {
        throw this._error;
    }
    get ui() {
        throw this._error;
    }
    get context() {
        throw this._error;
    }
}
/**
 * Container for common extension variables used throughout the AppService package. They must be initialized with registerAppServiceExtensionVariables
 */
exports.ext = new UninitializedExtensionVariables();
/**
 * Call this to register common variables used throughout the AppService package.
 */
function registerAppServiceExtensionVariables(extVars) {
    exports.ext = extVars;
    vscode_azureextensionui_1.registerUIExtensionVariables(extVars);
}
exports.registerAppServiceExtensionVariables = registerAppServiceExtensionVariables;
